app/src/main/java/com/realityexpander/tasky/agenda_feature/data/repositories/agendaRepository/agendaRepositoryImpls/AgendaRepositoryImpl.kt

```diff
@@ -98,63 +99,63 @@ class AgendaRepositoryImpl @Inject constructor(
 //                    }
 //                }
 
-                withContext(Dispatchers.IO) {
-
-                    // • Events - Insert fresh data locally in parallel
-                    eventRepository.clearEventsForDayLocally(dateTime)
-                    agenda?.events?.map { event ->
-                        supervisorScope {
-                            async {
-                                val upsertResult =
-                                    eventRepository.upsertEventLocally(event.toDomain())
-                                if (upsertResult is ResultUiText.Error) {
-                                    throw IllegalStateException(upsertResult.message.asStrOrNull())
+                    return withContext(Dispatchers.IO) {
```

Comment:
Since we are using Retrofit and Room, you shouldn´t need to change Dispatcher since those libs already do it

Reply:
I was getting a run-time error that the room database was being accessed on the main thread, so I had to use this to change the context. I thought room automatically changes the context but it doesnt in this case, so I force it here.

Reply:
Weird that would also be new to me. I never needed to do this or had this issue 🤔 
These are all suspend functions you expose in the DAO?

Reply:
No these particular functions are not suspend function, they are `@Transactions` due to needing the update/insert functionality.

Used here:

https://github.com/realityexpander/Tasky/blob/d4923498462d5fcc058b02ba9b2bfbafb61a0503/app/src/main/java/com/realityexpander/tasky/agenda_feature/data/repositories/eventRepository/eventRepositoryImpls/EventRepositoryImpl.kt#L117-L126

Defined Here:

https://github.com/realityexpander/Tasky/blob/6bed3b1e704962d153c761bb90842d7588eb38a0/app/src/main/java/com/realityexpander/tasky/agenda_feature/data/repositories/eventRepository/local/eventDao/eventDaoImpls/EventDaoImpl.kt#L54-L68

Is there a better way to approach this? I thought the `@Transaction` requires to be not susped functions. Im confised a bit here about the approach.

## FOLLOW UP QUESTION 3

Please advise

Reply:
Ah okay, if these can't be suspending the error makes sense, since then Room can't switch the dispatcher behind the scenes. I assume it's because Room needs to make sure they are executed on the same thread to avoid race conditions which it can't fully do if it's a suspend function.

Reply:
I refactored to use suspend for these funtions.